src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp

```diff
@@ -380,46 +380,48 @@ void ShenandoahOldGeneration::transition_to(State new_state) {
-//               |   |    |     |    BOOTSTRAP    |
-//               |   |    |     |                 |
-//               |   |    |     +-----------------+
-//               |   |    |       |
-//               |   |    |       | Continue Marking
-//               |   |    |       v
-//               |   |    |     +-----------------+     +----------------------+
-//               |   |    |     |    MARKING      | <-> |       YOUNG GC       |
-//               |   +----|-----|                 |     | (RSet Parses Region) |
-//               |        |     +-----------------+     +----------------------+
-//               |        |       |
-//               |        |       | Has Candidates
-//               |        |       v
-//               |        |     +-----------------+
-//               |        |     |    WAITING FOR  |
-//               +--------|---- |    EVACUATIONS  |
-//                        |     +-----------------+
-//                        |       |
-//                        |       | All Candidates are Pinned
-//                        |       v
-//                        |     +-----------------+
-//                        |     |    WAITING FOR  |
-//                        +-----|    FILLING      |
+//           +----------------- +-----------------+
+//           |   +------------> |      IDLE       |
+//           |   |   +--------> |                 |
+//           |   |   |          +-----------------+
+//           |   |   |            |
+//           |   |   |            | Begin Old Mark
+//           |   |   |            v
+//           |   |   |          +-----------------+     +--------------------+
+//           |   |   |          |     FILLING     | <-> |      YOUNG GC      |
+//           |   |   |    +---> |                 |     | (RSet Uses Bitmap) |
+//           |   |   |    |     +-----------------+     +--------------------+
+//           |   |   |    |       |
+//           |   |   |    |       | Reset Bitmap
+//           |   |   |    |       v
+//           |   |   |    |     +-----------------+
+//           |   |   |    |     |    BOOTSTRAP    |
+//           |   |   |    |     |                 |
+//           |   |   |    |     +-----------------+
+//           |   |   |    |       |
+//           |   |   |    |       | Continue Marking
+//           |   |   |    |       v
+//           |   |   |    |     +-----------------+     +----------------------+
+//           |   |   |    |     |    MARKING      | <-> |       YOUNG GC       |
+//           |   |   +----|-----|                 |     | (RSet Parses Region) |
+//           |   |        |     +-----------------+     +----------------------+
+//           |   |        |       |
+//           |   |        |       | Has Candidates
+//           |   |        |       v
+//           |   |        |     +-----------------+
+//           |   |        |     |    WAITING FOR  |
+//           |   +--------|---> |    EVACUATIONS  |
```

Comment:
This transition has arrows at both ends (IDLE and WAITING FOR EVACUATIONS).

Reply:
That is intentional, I shall also add an arrow at both ends for IDLE and WAITING FOR FILL. Global cycles do not progress the old generation through marking/bootstrapping so they are allowed to move the old generation from IDLE to one of these states at the end of global mark.

Reply:
I see, so a bidirectional arrow?

It would of course be ideal (in the fullness of time) if there were events labeling the transitions, even if this required a table for the event alphabet.